plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataRebuilder.java

```diff
@@ -0,0 +1,418 @@
+
+                @Override
+                public void end() {
+                  if (currentGroupId != null) {
+                    processMetadata(metadataMavenPath(currentGroupId, null, null), metadataBuilder.onExitGroupId());
+                  }
+                }
+              }
+          )
+      ).execute(sqlParams);
+    }
+
+    /**
+     * Method processing in separate TX/DB, performs writes (and mangles ThreadLocal DB, so caller should restore it).
+     * Handles A and bV on metadataBuilder.
+     */
+    private void rebuildAbV(final ORID bucket,
+                            final String groupId,
+                            final String artifactId,
+                            final Set<String> baseVersions)
+    {
+      metadataBuilder.onEnterArtifactId(artifactId);
+      for (String baseVersion : baseVersions) {
+        metadataBuilder.onEnterBaseVersion(baseVersion);
+        try (StorageTx tx = storageFacet.openTx()) {
+          final Iterable<Component> components = tx.findComponents(
+              "bucket = :bucket and group = :groupId and name = :artifactId and attributes.maven2.baseVersion = :baseVersion",
+              ImmutableMap
+                  .<String, Object>of("bucket", bucket, "groupId", groupId, "artifactId", artifactId, "baseVersion",
+                      baseVersion),
+              null,
+              "order by version asc");
+          for (Component component : components) {
+            final Iterable<Asset> assets = tx.browseAssets(component);
+            for (Asset asset : assets) {
+              final MavenPath mavenPath = mavenPathParser.parsePath(
+                  asset.formatAttributes().require(StorageFacet.P_PATH, String.class)
+              );
+              if (mavenPath.isSubordinate()) {
+                continue;
+              }
+              metadataBuilder.addArtifactVersion(mavenPath);
+              mayUpdateChecksum(asset, mavenPath, HashType.SHA1);
+              mayUpdateChecksum(asset, mavenPath, HashType.MD5);
+              if (mavenPath.getCoordinates() != null && mavenPath.getCoordinates().getExtension().equals("pom")) {
+                final Document pom = getModel(mavenPath);
+                if (pom != null) {
+                  final String packaging = getChildValue(pom, "packaging", "jar");
+                  if ("maven-plugin".equals(packaging)) {
+                    metadataBuilder.addPlugin(getPluginPrefix(mavenPath.locateMain("jar")), artifactId,
+                        getChildValue(pom, "name", null));
+                  }
+                }
+              }
+            }
+          }
+        }
+        processMetadata(metadataMavenPath(groupId, artifactId, baseVersion), metadataBuilder.onExitBaseVersion());
+      }
+      processMetadata(metadataMavenPath(groupId, artifactId, null), metadataBuilder.onExitArtifactId());
+    }
+
+    /**
+     * Verifies and may fix/create the broken/non-existent Maven hashes (.sha1/.md5 files).
```

Comment:
Missing checksums from Assets, yes, as that would mean our blobstore/storage/mavenFacet did not do it's job (have bug or db corrupted?). Missing _maven checksums_ OTOH are fixed/created as needed (as separate assets).